File: react/react/ios/App/App/WidgetBridgePlugin.swift
import Foundation
import UIKit
import Capacitor
import WidgetKit

/// Capacitor 6 only auto-registers plugins that ship as npm packages (via
/// capacitor.config.json's packageClassList). App-local plugins like
/// WidgetBridgePlugin must be registered explicitly — done here through the
/// bridge's `capacitorDidLoad()` hook. Main.storyboard's root view controller is
/// repointed from CAPBridgeViewController to this subclass.
class MainViewController: CAPBridgeViewController {
    override open func capacitorDidLoad() {
        bridge?.registerPluginInstance(WidgetBridgePlugin())
    }
}

/// Bridges the web app's widget snapshot into the shared App Group container so
/// the NetWorthWidget extension can read it, then refreshes the widget.
///
/// JS side: `registerPlugin('WidgetBridge')` → `WidgetBridge.publish({ payload })`
/// (see `src/utils/widgetData.ts`). The App Group id must match the capability
/// added to BOTH the app and the widget targets in Xcode.
@objc(WidgetBridgePlugin)
public class WidgetBridgePlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "WidgetBridgePlugin"
    public let jsName = "WidgetBridge"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "publish", returnType: CAPPluginReturnPromise)
    ]

    /// Must exactly match the App Group id configured on the app + widget targets.
    static let appGroupId = "group.com.mypokerface.app"
    static let snapshotKey = "widgetSnapshot"

    @objc func publish(_ call: CAPPluginCall) {
        guard let payload = call.getString("payload") else {
            call.reject("payload is required")
            return
        }
        guard let defaults = UserDefaults(suiteName: Self.appGroupId) else {
            call.reject("App Group \(Self.appGroupId) is not available")
            return
        }
        defaults.set(payload, forKey: Self.snapshotKey)
        if #available(iOS 14.0, *) {
            WidgetCenter.shared.reloadAllTimelines()
        }
        call.resolve()
    }
}

File: react/react/android/app/src/main/java/com/mypokerface/app/WidgetBridgePlugin.java
package com.mypokerface.app;

import android.content.Context;
import android.content.SharedPreferences;

import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

/**
 * Android side of the web app's home-screen widget bridge — the counterpart to
 * iOS's WidgetBridgePlugin.swift. The JS side
 * (src/utils/widgetData.ts → registerPlugin('WidgetBridge').publish({ payload }))
 * sends the same JSON snapshot on both platforms.
 *
 * Unlike iOS (where the widget is a separate process and needs an App Group
 * container), an Android app-widget runs in this same app package, so it can read
 * our own SharedPreferences directly — no shared container needed. We write the
 * payload, then poke AppWidgetManager so {@link NetWorthWidgetProvider} re-renders.
 *
 * Registered in MainActivity.onCreate (Capacitor 6 only auto-registers plugins
 * that ship as npm packages; app-local ones must be registered explicitly).
 */
@CapacitorPlugin(name = "WidgetBridge")
public class WidgetBridgePlugin extends Plugin {

    /** Shared with NetWorthWidgetProvider — both must agree on these names. */
    static final String PREFS_NAME = "net_worth_widget";
    static final String SNAPSHOT_KEY = "widgetSnapshot";

    @PluginMethod
    public void publish(PluginCall call) {
        String payload = call.getString("payload");
        if (payload == null) {
            call.reject("payload is required");
            return;
        }

        Context ctx = getContext().getApplicationContext();
        SharedPreferences prefs = ctx.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefs.edit().putString(SNAPSHOT_KEY, payload).apply();

        NetWorthWidgetProvider.requestUpdate(ctx);
        call.resolve();
    }
}
